crates/jxl-frame/src/data/noise.rs

```diff
@@ -17,3 +26,190 @@ impl<Ctx> jxl_bitstream::Bundle<Ctx> for NoiseParameters {
         Ok(Self { lut })
     }
 }
+
+#[inline]
+fn rng_seed0(visible_frames: u64, invisible_frames: u64) -> u64 {
+    (visible_frames << 32) + invisible_frames
+}
+
+#[inline]
+fn rng_seed1(x0: usize, y0: usize) -> u64 {
+    ((x0 as u64) << 32) + y0 as u64
+}
+
+pub fn init_noise(
+    visible_frames: u64,
+    invisible_frames: u64,
+    header: &FrameHeader,
+) -> [SimpleGrid<f32>; 3] {
+    let width = header.width as usize;
+    let height = header.height as usize;
+    let group_dim = header.group_dim() as usize;
+    let seed0 = rng_seed0(visible_frames, invisible_frames);
+
+    let mut noise_buffer: [SimpleGrid<f32>; 3] = [
+        SimpleGrid::new(width, height),
+        SimpleGrid::new(width, height),
+        SimpleGrid::new(width, height),
+    ];
+
+    let groups_per_row = (width + group_dim - 1) / group_dim;
+    let groups_num = groups_per_row * ((height + group_dim - 1) / group_dim);
+
+    for group_idx in 0..groups_num {
+        // TODO: It may be necessary to multiply by `upsamping`
+        let group_x = group_idx % groups_per_row;
+        let group_y = group_idx / groups_per_row;
+        // TODO: additional groups for upsampling may be needed
+        let x0 = group_x * group_dim;
+        let y0 = group_y * group_dim;
+        init_noise_group(seed0, &mut noise_buffer, x0, y0, width, height, group_dim);
+    }
+
+    let mut laplacian = [[0f32; 5]; 5];
+    (0..5).for_each(|y| {
+        (0..5).for_each(|x| laplacian[y][x] = if x == 2 && y == 2 { -3.84 } else { 0.16 });
+    });
+
+    let mut convolved: [SimpleGrid<f32>; 3] = [
+        SimpleGrid::new(width, height),
+        SimpleGrid::new(width, height),
+        SimpleGrid::new(width, height),
+    ];
+
+    for c in 0..3 {
+        let cbuffer = convolved[c].buf_mut();
+        let nbuffer = noise_buffer[c].buf();
+        (0..height).for_each(|y| {
+            (0..width).for_each(|x| {
+                (0..5).for_each(|iy| {
+                    (0..5).for_each(|ix| {
+                        let cy = mirror((y + iy) as i32 - 2, height as i32);
+                        let cx = mirror((x + ix) as i32 - 2, width as i32);
+                        cbuffer[y * width + x] += nbuffer[cy * width + cx] * laplacian[iy][ix];
+                    });
+                });
```

Comment:
In my experience, making grids a little larger and copying edge samples was faster than explicit `Mirror` calls.